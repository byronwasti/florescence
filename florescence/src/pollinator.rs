use crate::reality_token::RealityToken;
use treeclocks::{EventTree, ItcMap};

mod identity_map;

pub use identity_map::IdentityMap;

pub struct PeerInfo {}

pub trait Pollinator {
    type A: PollinatorInner + Sized;
    fn from_conn() -> (Self, Self::A)
    where
        Self: Sized;
}

pub trait PollinatorInner {}

#[allow(dead_code)]
pub(crate) struct PollinatorCore {
    timestamp: EventTree,
    reality_token: RealityToken,
    logic: Box<dyn PollinatorInner>,
    peer_info: ItcMap<PeerInfo>,
}
