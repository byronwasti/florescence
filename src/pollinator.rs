use crate::*;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use treeclocks::{EventTree, ItcMap};

mod identity_map;

pub use identity_map::IdentityMap;

#[derive(Debug, Serialize, Deserialize)]
pub struct RealityToken(u64);
pub struct PeerInfo {}

pub trait Pollinator {
    type A: PollinatorInner + Sized;
    fn from_conn() -> (Self, Self::A)
    where
        Self: Sized;
}

pub trait PollinatorInner {}

pub(crate) struct PollinatorCore {
    timestamp: EventTree,
    reality_token: RealityToken,
    logic: Box<dyn PollinatorInner>,
    peerInfo: ItcMap<PeerInfo>,
}

#[derive(Debug, Serialize, Deserialize)]
pub(crate) struct UpdatePacket {
    inner: u32,
}
