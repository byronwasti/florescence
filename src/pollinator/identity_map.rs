use super::{Pollinator, PollinatorInner};

pub struct IdentityMap<T> {
    t: T,
}

impl<T> IdentityMap<T> {
    fn set(&mut self, value: T) {
        todo!()
    }

    fn get(&self) -> T {
        todo!()
    }

    fn fold<B, F>(&self, init: B, f: F) -> B
    where
        F: FnMut(B, T) -> B,
    {
        todo!()
    }

    fn apply<F>(&self, f: F) -> T
    where
        F: FnMut(T) -> T,
    {
        todo!()
    }
}

impl<T> PollinatorInner for IdentityMap<T> {}

impl<T> Pollinator for IdentityMap<T> {
    type A = IdentityMap<T>;

    fn from_conn() -> (Self, Self::A) {
        todo!()
    }
}
