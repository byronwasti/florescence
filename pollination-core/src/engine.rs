use crate::message::PollinationMessage;
use serde::{Deserialize, Serialize};
use std::{fmt, future::Future, hash::Hash};
use tokio::sync::mpsc::{Receiver, Sender};

const DEFAULT_CHANNEL_SIZE: usize = 10;

#[cfg(feature = "axum")]
pub mod axum;

pub trait Engine: 'static {
    type Addr: Clone + Serialize + for<'de> Deserialize<'de> + Hash + fmt::Display + Send;
    type Error: std::error::Error + 'static;

    fn run_background(
        self,
    ) -> impl Future<
        Output = Result<(Sender<EngineRequest<Self::Addr>>, Receiver<EngineEvent>), Self::Error>,
    > + Send;
}

pub struct EngineRequest<A> {
    pub pollination_msg: PollinationMessage,
    pub addr: A,
    pub tx: Sender<PollinationMessage>,
}

pub struct EngineEvent {
    pub pollination_msg: PollinationMessage,
    pub tx: Sender<PollinationMessage>,
}
