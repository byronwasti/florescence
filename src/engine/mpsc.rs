use crate::constants::MPSC_CHANNEL_SIZE;
use crate::engine::Engine;
use serde::{Deserialize, Serialize};
use std::sync::{Arc, RwLock};
use tokio::sync::mpsc::{self, Receiver, Sender};

type FloraConn<T> = (Sender<T>, Receiver<T>);
type EngineConn<T> = Sender<FloraConn<T>>;

pub type World<T> = Arc<RwLock<Vec<EngineConn<T>>>>;

pub fn new_world<T>() -> Arc<RwLock<Vec<EngineConn<T>>>> {
    Arc::new(RwLock::new(vec![]))
}

pub struct MpscEngine<T> {
    world_view: World<T>,
    own_idx: usize,
    receiver: Receiver<FloraConn<T>>,
}

impl<T> MpscEngine<T> {
    pub fn new(world_view: World<T>) -> Self {
        let (own_idx, rx) = {
            let mut world = world_view.write().expect("poisoned lock");
            let (tx, rx) = mpsc::channel(MPSC_CHANNEL_SIZE);
            world.push(tx);
            (world.len() - 1, rx)
        };

        Self {
            own_idx,
            world_view,
            receiver: rx,
        }
    }
}

impl<T> Engine<T> for MpscEngine<T>
where
    T: Serialize + for<'a> Deserialize<'a> + Clone + Send,
{
    type Addr = usize;

    fn start(&mut self) {
        // Nothing to do.
    }

    fn addr(&self) -> &Self::Addr {
        &self.own_idx
    }

    async fn create_conn(&mut self, addr: usize) -> (Sender<T>, Receiver<T>) {
        let peer_tx = {
            let world = self.world_view.read().expect("poisoned lock");

            world[addr].clone()
        };

        let (tx0, rx0) = mpsc::channel(MPSC_CHANNEL_SIZE);
        let (tx1, rx1) = mpsc::channel(MPSC_CHANNEL_SIZE);

        peer_tx.send((tx0, rx1)).await;

        (tx1, rx0)
    }

    async fn get_new_conn(&mut self) -> Option<(Sender<T>, Receiver<T>)> {
        self.receiver.recv().await
    }
}
