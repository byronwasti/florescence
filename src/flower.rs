use crate::pollinator::{RealityToken, Pollinator};
use crate::engine::{Engine, EngineConnection};
use std::collections::{HashMap, HashSet};
use treeclocks::{EventTree, ItcMap, itc_map::UpdatePacket};
use std::sync::Arc;
use std::time::Duration;

struct Topic(String);

struct FlowerCore<I, A> {
    id: I,
    addr: A,
    peer_info: HashMap<I, PeerInfo<I, A>>,
}

struct PeerInfo<I, A> {
    id: I,
    addr: A,
    topics: HashSet<Topic>,
}


pub struct Flower {
    conn: EngineConnection,
}

impl Flower {
    /*
    pub fn builder<I, E>() -> FlowerBuilder<I, E> {
        FlowerBuilder::default()
    }
    */

    pub fn pollinator<P: Pollinator + 'static>(&self, interval: Duration) -> P {
        let (pollinator, inner) = P::from_conn(EngineConnection {});
        pollinator
    }
}

/*
pub struct FlowerBuilder<I, E> {
    id: Option<I>,
    engine: Option<E>,
}

impl<T, I, E: Engine<T, I>> FlowerBuilder<I, E> {
    pub fn id(mut self, id: I) -> Self {
        self.id = Some(id);
        self
    }

    pub fn engine(mut self, engine: E) -> Self {
        self.engine = Some(engine);
        self
    }

    pub async fn bloom(mut self) -> Result<Flower, FlowerError> {

        // TODO: Kick off background task
        //let conn = self.engine.unwrap().run().await;
        //Ok(Flower { conn })
        todo!()
    }
}

impl<I, E> Default for FlowerBuilder<I, E> {
    fn default() -> FlowerBuilder<I, E> {
        FlowerBuilder {
            id: None,
            engine: None,
        }
    }
}
*/

pub enum FlowerError {}
