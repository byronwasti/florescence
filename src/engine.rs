use crate::pollinator::PollinatorCore;
mod tonic_engine;
pub use tonic_engine::TonicEngine;
use tokio::sync::mpsc::{UnboundedSender, UnboundedReceiver};
use crate::message::PollinationMessage;
use serde::{Serialize, Deserialize};

pub trait Engine<T, I> {
    type Addr;

    /// Must be non-blocking
    fn start(&mut self);

    fn addr(&self) -> &Self::Addr;

    fn create_conn(&mut self, addr: Self::Addr) -> Connection<T, I, Self::Addr>;

    fn get_new_conns(&mut self) -> Vec<Connection<T, I, Self::Addr>>;
}

pub(crate) struct EngineCore<E, T, I, A> {
    engine: E,
    pollinators: Vec<PollinatorCore>,
    connections: Vec<Connection<T, I, A>>,
}


pub struct Connection<T, I, A> {
    pub addr: A,
    pub tx: UnboundedSender<PollinationMessage<T, I, A>>,
    pub rx: UnboundedReceiver<PollinationMessage<T, I, A>>,
}

// TODO: Remove
pub struct EngineConnection;
