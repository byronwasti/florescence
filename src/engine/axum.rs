use crate::{
    ds::WalkieTalkie,
    message::PollinationMessage,
    serialization::{deserialize, serialize},
};
use axum::{
    Router,
    body::Bytes,
    extract::State,
    http::StatusCode,
    response::IntoResponse,
    routing::{get, post},
};
use serde::{Deserialize, Serialize};
use std::{net::SocketAddr, sync::Arc};
use thiserror::Error;
use tokio::sync::mpsc::{Receiver, Sender, channel};
use url::Url;

use super::{DEFAULT_CHANNEL_SIZE, Engine, EngineEvent, EngineRequest};

pub struct AxumEngine {
    socket_addr: SocketAddr,
}

impl AxumEngine {
    fn new(socket_addr: SocketAddr) -> Self {
        Self { socket_addr }
    }
}

impl Engine for AxumEngine {
    type Addr = Url;
    type Error = AxumEngineError;

    async fn run(
        self,
        addr: Self::Addr,
    ) -> Result<WalkieTalkie<EngineRequest<Self::Addr>, EngineEvent>, Self::Error> {
        let (wt0, wt1) = WalkieTalkie::pair();

        let (tx, rx) = wt1.split();

        tokio::spawn(sender_task(rx));

        let state = Arc::new(AppState { tx, addr });

        let app = Router::new()
            .route("/", post(handle_message))
            .with_state(state);

        let listener = tokio::net::TcpListener::bind(self.socket_addr).await?;
        tokio::spawn(async move {
            let res = axum::serve(listener, app).await;
            if let Err(err) = res {
                error!("Error running Axum: {err:?}");
            }
        });

        Ok(wt0)
    }
}

async fn handle_message(State(state): State<Arc<AppState>>, bytes: Bytes) -> impl IntoResponse {
    match handle_message_inner(&state.tx, bytes).await {
        Ok(msg) => (StatusCode::OK, msg),
        Err(err) => (StatusCode::INTERNAL_SERVER_ERROR, Bytes::new()),
    }
}

async fn handle_message_inner(
    tx: &Sender<EngineEvent>,
    bytes: Bytes,
) -> Result<Bytes, AxumEngineError> {
    let pollination_msg: PollinationMessage = deserialize(bytes.to_vec())?;
    let (res_tx, mut rx) = channel(DEFAULT_CHANNEL_SIZE);
    tx.send(EngineEvent {
        pollination_msg,
        tx: res_tx,
    });

    if let Some(res) = rx.recv().await {
        Ok(serialize(res)?.into())
    } else {
        Ok(Bytes::new())
    }
}

async fn sender_task(mut rx: Receiver<EngineRequest<Url>>) {
    loop {
        match rx.recv().await {
            Some(req) => {
                let EngineRequest {
                    pollination_msg,
                    addr,
                    tx,
                } = req;

                match send_and_recv(addr, pollination_msg).await {
                    Ok(res) => {
                        tx.send(res);
                    }
                    Err(err) => {
                        error!("Error sending request: {err}");
                    }
                }
            }
            None => {
                info!("Channel closed")
            }
        }
    }
}

async fn send_and_recv(
    addr: Url,
    pollination_msg: PollinationMessage,
) -> Result<PollinationMessage, AxumEngineError> {
    let client = reqwest::Client::new();
    let res = client
        .post(addr)
        .body(serialize(pollination_msg)?)
        .send()
        .await?;

    let bytes = res.bytes().await?.to_vec();
    let pollination_msg = crate::serialization::deserialize(bytes)?;

    Ok(pollination_msg)
}

#[derive(Debug, Error)]
pub enum AxumEngineError {
    #[error("Axum error: {0}")]
    Axum(#[from] axum::Error),

    #[error("StdIO error: {0}")]
    StdIo(#[from] std::io::Error),

    #[error("Deserialize error: {0}")]
    Deserialize(#[from] crate::serialization::DeserializeError),

    #[error("Serialize error: {0}")]
    Serialize(#[from] crate::serialization::SerializeError),

    #[error("Reqwest error: {0}")]
    Reqwest(#[from] reqwest::Error),
}

struct AppState {
    tx: Sender<EngineEvent>,
    addr: Url,
}

#[derive(Serialize, Deserialize)]
struct AxumMessage<A> {
    return_addr: A,
    pollination_msg: PollinationMessage,
}

impl<A> AxumMessage<A> {
    fn new(return_addr: A, pollination_msg: PollinationMessage) -> Self {
        Self {
            return_addr,
            pollination_msg,
        }
    }
}
