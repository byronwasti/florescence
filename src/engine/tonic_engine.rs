use std::pin::Pin;
use tokio_stream::{wrappers::UnboundedReceiverStream, Stream, StreamExt};
use tonic::{Request, Response, Status, Streaming, transport::Server};
use tokio::sync::mpsc::{self, UnboundedSender, UnboundedReceiver};
use crate::message::PollinationMessage;
use uuid::Uuid;
use http::Uri;
use std::net::SocketAddr;
use tracing::debug;
use serde::{Serialize, Deserialize};

mod codec;
mod rpc;

use super::*;
use rpc::{
    TonicReqWrapper,
    gossip_server::{Gossip, GossipServer},
    gossip_client::{GossipClient},
};


/// Streaming RPC via Tonic library
pub struct TonicEngine{
    socket_addr: SocketAddr,
    addr: Uri,
    tx: Option<UnboundedSender<EngMessage<Uri>>>,
}

impl TonicEngine {
    /// `socket_addr` is the connection on the server side
    /// `addr` is the address clients use
    fn new(socket_addr: SocketAddr, addr: Uri) -> Self {
        Self {
            socket_addr,
            addr,
            tx: None,
        }
    }
}

impl Engine for TonicEngine
{
    type Addr = Uri;

    fn addr(&self) -> &Self::Addr {
        &self.addr
    }

    fn remove_conn(&mut self, addr: Self::Addr) -> impl std::future::Future<Output=()> + Send {
        async {}
    }

    fn new_conn<T, I, A>(&mut self, addr: Uri) -> impl std::future::Future<Output=(UnboundedSender<PollinationMessage<T, I, A>>, UnboundedReceiver<PollinationMessage<T, I, A>>)> + Send
    where T: for<'a> Deserialize<'a> + Serialize + Send + Sync + 'static,
          I: for<'a> Deserialize<'a> + Serialize + Send + Sync + 'static,
          A: for<'a> Deserialize<'a> + Serialize + Send + Sync + 'static,
    {
        let (tx0, rx0) = mpsc::unbounded_channel();
        let (tx1, rx1) = mpsc::unbounded_channel();
        tokio::task::spawn(async move {
            // TODO: Deal with failure
            let mut client = GossipClient::connect(addr.clone()).await.unwrap();
            
            let in_stream = UnboundedReceiverStream::new(rx0)
                .map(|x: PollinationMessage<_, _, _>| {
                    TonicReqWrapper {
                        raw: bincode::serde::encode_to_vec(x, bincode::config::standard()).expect("Unable to serialize message"),
                    }
                });
            // TODO: Deal with failure
            let mut res = client.gossip(in_stream)
                .await
                .unwrap();

            let mut out_stream = res.into_inner();

            loop {
                match out_stream.next().await {
                    Some(Ok(val)) => {
                        if let Ok((val, _)) = bincode::serde::decode_from_slice(&val.raw, bincode::config::standard()) {
                            if let Err(err) = tx1.send(val) {
                                debug!("Internal mpsc errored: {err}");
                                break
                            }
                        } else {
                            break
                        }
                    }
                    Some(Err(err)) => {
                        debug!("Receiving stream errored: {err}");
                        break
                    }
                    None => {
                        debug!("Receiving stream empty.");
                        break
                    }

                }
            }
            /*
            while let Some(rec) = out_stream.next().await {
                if let Err(_) = tx1.send(Some(rec)) {
                    break
                }
            }
            */

            //tx1.send(EngMessage::Terminated(addr))
        });

        std::future::ready((tx0, rx1))
    }

    fn start(&mut self, tx: UnboundedSender<EngMessage<Uri>>) -> impl std::future::Future<Output=()> + Send
    {
        self.tx = Some(tx.clone());
        run(self.socket_addr.clone(), tx)
    }
}

async fn run(socket_addr: SocketAddr, tx: UnboundedSender<EngMessage<Uri>>) {
    let gossiper = Handler::new(tx);

    tokio::task::spawn(async move {
        Server::builder()
            .add_service(GossipServer::new(gossiper))
            .serve(socket_addr)
            .await
            .expect("TonicRPC internal failure.")
    });
}

struct Handler {
    tx: UnboundedSender<EngMessage<Uri>>,
}

impl Handler {
    pub fn new(tx: UnboundedSender<EngMessage<Uri>>) -> Self {
        Self {
            tx
        }
    }
}

type ResponseStream = Pin<Box<dyn Stream<Item = Result<TonicReqWrapper, Status>> + Send>>;

#[tonic::async_trait]
impl Gossip for Handler {
    type GossipStream = ResponseStream;
    async fn gossip(
        &self,
        mut request: Request<Streaming<TonicReqWrapper>>,
    ) -> Result<Response<ResponseStream>, Status> {

        // TODO: This must be coordinated with the EngineCore
        let (tx0, rx0) = mpsc::unbounded_channel();
        let (tx1, rx1) = mpsc::unbounded_channel();

        let mut in_stream = request.into_inner();

        tokio::spawn(async move {
            while let res = in_stream.message().await {
                match res {
                    Ok(None) => {
                        // Stream is closed by peer
                        todo!()
                    }
                    Ok(v) => {
                        tx1.send(v);
                    }
                    Err(err) => {
                    }
                }
            }
        });

        let out_stream = UnboundedReceiverStream::new(rx0);
        Ok(Response::new(
                Box::pin(out_stream) as Self::GossipStream
        ))
    }
}

