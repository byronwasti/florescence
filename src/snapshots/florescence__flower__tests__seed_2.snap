---
source: src/flower.rs
assertion_line: 744
expression: "(f0.nucleus, f1.nucleus)"
---
(
    Nucleus {
        uuid: 00000000-0000-0000-0000-000000000001,
        propagativity: Propagating(
            SubTree(
                Zero,
                One,
            ),
        ),
        reality_token: RealityToken(
            3,
        ),
        core_map: ItcMap {
            timestamp: Leaf(
                3,
            ),
            data: [
                Some(
                    (
                        SubTree(
                            One,
                            Zero,
                        ),
                        PeerInfo {
                            uuid: 00000000-0000-0000-0000-000000000002,
                            addr: Some(
                                1,
                            ),
                            status: Healthy,
                        },
                    ),
                ),
                Some(
                    (
                        SubTree(
                            Zero,
                            One,
                        ),
                        PeerInfo {
                            uuid: 00000000-0000-0000-0000-000000000001,
                            addr: Some(
                                0,
                            ),
                            status: Healthy,
                        },
                    ),
                ),
            ],
            index: SubTree(
                Leaf(
                    0,
                ),
                Leaf(
                    1,
                ),
            ),
        },
        own_info: PeerInfo {
            uuid: 00000000-0000-0000-0000-000000000001,
            addr: Some(
                0,
            ),
            status: Healthy,
        },
    },
    Nucleus {
        uuid: 00000000-0000-0000-0000-000000000002,
        propagativity: Propagating(
            SubTree(
                One,
                Zero,
            ),
        ),
        reality_token: RealityToken(
            3,
        ),
        core_map: ItcMap {
            timestamp: Leaf(
                3,
            ),
            data: [
                Some(
                    (
                        SubTree(
                            Zero,
                            One,
                        ),
                        PeerInfo {
                            uuid: 00000000-0000-0000-0000-000000000001,
                            addr: Some(
                                0,
                            ),
                            status: Healthy,
                        },
                    ),
                ),
                Some(
                    (
                        SubTree(
                            One,
                            Zero,
                        ),
                        PeerInfo {
                            uuid: 00000000-0000-0000-0000-000000000002,
                            addr: Some(
                                1,
                            ),
                            status: Healthy,
                        },
                    ),
                ),
            ],
            index: SubTree(
                Leaf(
                    1,
                ),
                Leaf(
                    0,
                ),
            ),
        },
        own_info: PeerInfo {
            uuid: 00000000-0000-0000-0000-000000000002,
            addr: Some(
                1,
            ),
            status: Healthy,
        },
    },
)
